platform-sdk/docs/proposals/001-tss-rosters-design-proposal/001-tss-rosters-design-proposal.md

```diff
@@ -0,0 +1,265 @@
+sets the roster) with clear responsibility between the app and the platform.
+
+Some edge cases worth considering include:
+
+1. multiple concurrent roster submission - has a potential for introducing race conditions. We will prevent this by
+   guaranteeing immutability on rosters.
+2. Size control on the map of rosters - we certainly don’t want this map to grow infinitely so insertion of rosters will
+   be controlled, with clear rules for removal of unused rosters (e.g. the acceptance of a new candidate roster
+   invalidates and removes existing one).
+
+###### Roster API - Implementation
+
+Implementation of some new TSS components has been proposed.
+See [link](https://www.notion.so/TSS-Platform-Architecture-04b15df371ba4b1d848360542e05a030?pvs=21).
+
+It is **important** to note that before the TSS protocol is implemented and launched, the Platform will adopt a CR
+during any software upgrade if the CR is present in the state. Only after the TSS protocol is launched will we be able
+to add an extra condition to verify that the CR is complete, meaning that it has enough TSS key material and votes
+accumulated.
+
+### Core Behaviors, in summary
+
+- Roster Creation: App will create a candidate roster from the Candidate Address Book (CAB) and set it in the state.
+- Roster Submission: App will trigger roster adoption by setting the candidateRosterHash field in the PlatformState.
+- Roster Adoption: The platform will adopt the last submitted candidate roster when it is ready, based on TSS key
+  material and vote accumulation.
+- Roster Replacement: If a new CAB is submitted before the previous one is adopted, the new candidate roster will
+  replace the old one.
+
+### Public API
+
+Platform API: A new method will be added to the platform API to allow the App submit a candidate roster:
+
+```java
+//in State
+
+void submitCandidateRoster(@NonNull final Roster candidateRoster);
+```
+
+### Configuration
+
+The config.txt file will no longer be used for storing the address book in existing networks. It will only be used for
+the genesis of new networks.
+
+### Data storage
+
+All the data related to active and candidate roster, including the key material being accumulated via TSS messages, will
+always be stored in the State. There will not be any other separate artifacts stored elsewhere (e.g. directly on disk.).
+The only artifact that will continue to be stored on disk separately from the State is the config.txt, which, as
+explained in *Bootstrapping Genesis for a brand new network* section below, will only ever be used once during a genesis
+of a brand new network. The config.txt file, or at least its part that describes the address book components, will never
+be used in the life cycle of an existing network after that.
+
+### Startup procedure (pseudo-code)
+
+```java
+if(a State
+exists on
+disk){
+
+loadStateFromDisk();
+}else{
```

Comment:
I suggest to preserve the original lines indentation so that this code is readable.

Reply:
I agree, the current formatting makes this very difficult to read

Reply:
The formatting got lost after copy/paste from notion. Fixed